### DISCUSSION TITLE
Inconsistent Data Structure in Hooks

### DISCUSSION THREAD

**Discussion by sarthakgaur** [directus/directus#11196] [Archive]
In the new hooks there are some inconsistencies:

1. In the `items.update` event in `filter` and `action`, the meta object contains `keys`. Shouldn't this be just `key`?
2. In the `items.delete` event in `filter`, the `payload` is the key. I think the key should be in the `meta` object.

**Comment by joselcvarela** [comment]
Regarding the keys we are aware of this behaviour. Also there's already a discussion about it here https://github.com/directus/directus/discussions/10981

About the `key` in `meta` it makes sense since we also have them for `update`.

That being said, do you want to try a PR? 🙂

**Comment by rijkvanzanten** [comment]
> In the items.delete event in filter, the payload is the key. I think the key should be in the meta object.

This makes sense, as that's the only payload that exists for a delete operation. You're deleting items by primary key 👍🏻

**Reply by sarthakgaur** [reply]
That's fine. But, if you also make it available in the `meta` object, it would be consistent with all the hooks.

**Comment by nickrum** [comment]
> In the `items.update` event in `filter` and `action`, the meta object contains `keys`. Shouldn't this be just `key`?

If you update multiple items at once, a single action or filter will be triggered containing the keys of the updated items. If you create multiple items at once, an action or filter will be triggered for each newly created item and the action will contain the key of the new item. That's why the meta object for `items.update` contains a `keys` array and the meta object for `items.create` contains a single `key`.

**Reply by joselcvarela** [reply]
I understand what you are saying Nicola, although this should be transparent to the developer.
If we are asking to create/update/delete multiple items, as developer, I expect on hook to have a `keys` property.
On the other and, if we are asking to create/update/delete a single item, I expect on hook to have a `keys` with single item.
This is all about consistency, because we could also use `one_key`, `many_keys`, and so on. 🤔

**Reply by rijkvanzanten** [reply]
You're right from a DX perspective, but it is interesting from a tech perspective on the core side of things. Most "many" operations loop over the payload, and then call the "one" operation sequentially (as permissions checks etc have to be done) 🤔

**Reply by sarthakgaur** [reply]
@nickrum 

> If you update multiple items at once, a single action or filter will be triggered containing the keys of the updated items.

I didn't know about this behavior. I think most people will assume that the hooks will be triggered for every update. There are two things that can be done here:

1. Make the update hook trigger for every item update.
2. Remove the `key` property in `meta` in every other hook, and make it `keys` with just one item.

I have another design question here:

In the current design, if we update multiple items, a single `filter` and `action` hook will be triggered. Shouldn't this behavior be replicated for rest of the events like create and delete?

**Reply by rijkvanzanten** [reply]
The difference with update vs the others is that you update multiple records to the _same_ value. That's different from creates, as every row in a multiple-create will have different value

**Reply by sarthakgaur** [reply]
@rijkvanzanten Are you referring to the batch edit feature? If so, I think that's a special case. In most of the cases an update will only be performed on a single item. I think triggering the update hook on every item update, and using just `key` in the `meta` object looks better.
